[Xcode] Change header postprocessing logic to better reflect when it's needed internally

emw-apple opened the pull request:
#### 6e4108c84c0445fce64c1cccabd28635b62e90d6
<pre>
[Xcode] Change header postprocessing logic to better reflect when it&apos;s needed internally
<a href="https://bugs.webkit.org/show_bug.cgi?id=251818">https://bugs.webkit.org/show_bug.cgi?id=251818</a>
rdar://105031778

Reviewed by Alexey Proskuryakov.

WebKit currently assumes that API_AVAILABLE annotations will only use a
major OS version. We have to remember to bump this number annually, and
it prevents us from using minor versions in WebKit&apos;s API/SPI headers to
more accurately reflect availability.

Rather than make WebKit&apos;s internal configuration logic aware of all
software update versions (which is prohibitively difficult due to
xcconfig&apos;s limitations), introduce a new policy for when to enable
API_AVAILABLE annotations:

- Annotations are OFF in debug/release builds.
- Annotations are ON in production builds, unless the build has staged
  frameworks (i.e. it is a downlevel or STP build).

Internally, the production build environment may modify this policy by
overriding the WK_FRAMEWORK_HEADER_POSTPROCESSING_DISABLED setting.

* Source/WebKit/Configurations/WebKit.xcconfig:

Canonical link: <a href="https://commits.webkit.org/260346@main">https://commits.webkit.org/260346@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73af1fd96146b70ab142612a693d2181b1a1440e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108082 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17151 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40950 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117206 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18507 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8447 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100285 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113850 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/97175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/41884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/95859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28813 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/10035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/30161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10756 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/16183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49752 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7158 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12337 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov reviewed (approved):
Looks good, except that I don't understand why this is WebKit only. Don't we need this at least in JavaScriptCore, if not in WebKitLegacy too?

emw-apple committed:
[Xcode] Change header postprocessing logic to better reflect when it's needed internally
https://bugs.webkit.org/show_bug.cgi?id=251818
rdar://105031778

Reviewed by Alexey Proskuryakov.

WebKit currently assumes that API_AVAILABLE annotations will only use a
major OS version. We have to remember to bump this number annually, and
it prevents us from using minor versions in WebKit's API/SPI headers to
more accurately reflect availability.

Rather than make WebKit's internal configuration logic aware of all
software update versions (which is prohibitively difficult due to
xcconfig's limitations), introduce a new policy for when to enable
API_AVAILABLE annotations:

- Annotations are OFF in debug/release builds.
- Annotations are ON in production builds, unless the build has staged
  frameworks (i.e. it is a downlevel or STP build).

Internally, the production build environment may modify this policy by
overriding the WK_FRAMEWORK_HEADER_POSTPROCESSING_DISABLED setting.

* Source/WebKit/Configurations/WebKit.xcconfig:

Canonical link: https://commits.webkit.org/260346@main
